server/backend/testspg/router/test_router_application.py

```diff
@@ -244,4 +257,243 @@ def test_fam_application_endpoints_invlid_path_application_id_type(
         headers=jwt_utils.headers(token)
     )
     assert response.status_code == HTTPStatus.UNPROCESSABLE_ENTITY
-    assert invalid_path_router_msg in response.text
\ No newline at end of file
+    assert invalid_path_router_msg in response.text
+
+
+# ---------------- Test get application user role assignment filtering scenarios ------------ #
+
+def test_get_application_user_role_assignments_filtering_for_delegated_admin(
+    test_client_fixture,
+    test_rsa_key,
+    create_test_user_role_assignments,
+    fom_dev_access_admin_token,
+):
+    """
+    This test first create various user role assigments at FOM_DEV (IDIR/BCEID,
+    business organizations and forest client...) and verify different
+    visibility scenarios based on requester's admin level and
+    delegated admin privileges.
+    """
```

Comment:
Hi Ian, thanks so much for writing this test, I totally get the complexity. Just thinking of the maintainability and for easy understanding, maybe we could think of separate this one big test case?

Like we can do three different tests:

- Verify fom dev admin can get all user role assignments (we already have the test case `test_get_fam_application_user_role_assignment_concrete_role` and `test_get_fam_application_user_role_assignment_abstract_role`, so could also skip this one)

- Verify IDIR delegate admin can get all user role assignment with the privilege roles
    I got what you're trying to verify here, so basically we check:
    - response data total length 
    - response only has user role assignments with fom_reviewer and fom_submitter_00001018 role
    - response data length of idir user (with abstract role and concrete role)
    - response data length of bceid user (with abstract role and concrete role)
  
    I'm thinking if we can check everything based on one list instead of multiple ones. I found it's a little bit hard to follow when we check the length based on different lists. And also if we can add simple comment of what we're trying to verify for the above 4 cases?
    ```
     access_grants = [
        ACCESS_GRANT_FOM_DEV_CR_IDIR,
        ACCESS_GRANT_FOM_DEV_CR_BCEID_L3T,
        ACCESS_GRANT_FOM_DEV_AR_00001018_IDIR,          
        ACCESS_GRANT_FOM_DEV_AR_00000001_IDIR,
        ACCESS_GRANT_FOM_DEV_AR_00001018_BCEID_L3T
    ]
    ```
- Verify BCEID delegate admin can get all user role assignment with the privilege roles and from same organization
    I'm thinking maybe we can just create two lists as below, and we should still be able to verify
    - response data total length == len(access_grants_able_to_see)
    - response only has user role assignments with fom_reviewer and fom_submitter_00001018 role
    - response data length of idir user (==0)
    - response data length of bceid user == len(access_grants_able_to_see)
    - response data should all come from access_grants_able_to_see list
    ```
     access_grants_able_to_see = [
        ACCESS_GRANT_FOM_DEV_CR_BCEID_L3T,
        ACCESS_GRANT_FOM_DEV_AR_00001018_BCEID_L3T
    ]
    access_grants_not_able_to_see = [
        ACCESS_GRANT_FOM_DEV_CR_IDIR,
        ACCESS_GRANT_FOM_DEV_CR_BCEID_L4T, 
        ACCESS_GRANT_FOM_DEV_AR_00001018_IDIR,
        ACCESS_GRANT_FOM_DEV_AR_00001018_BCEID_L4T,
        ACCESS_GRANT_FOM_DEV_AR_00000001_BCEID_L3T
    ]
    ```

We can touch base tomorrow, this is a complicated one, just thinking if we can make it simpler to understand, thanks so much!!

Reply:
not sure how you like the idea of two lists`access_grants_able_to_see` and `access_grants_not_able_to_see`, maybe for the IDIR case we can also use these two lists, and verify the response data are all from the `access_grants_able_to_see`, maybe that's easier?

Reply:
No worry, Initially I was thinking about several tests but then decided to put them into one. I was hoping the added comments could be giving reader more understanding, but if it is still not clear for your to read, I can look at your suggestion and give it some changes.

Reply:
Hi @MCatherine1994 , you have some time?

Reply:
Hi Catherine, I split the test with some adjustment. Hopefully they are more clear to read now.
Thanks!